Crashsite cutscene

Zahir-Khan opened the pull request:
Added a cutscene for crashsite explaining spawn and outposts.

Zahir-Khan committed:
Added cutscene outpost

Zahir-Khan committed:
Added cutscene outpost

Zahir-Khan committed:
crashsite cutscene strings

Zahir-Khan committed:
crashsite cutscene strings

Zahir-Khan committed:
Create cutscene.lua

Zahir-Khan committed:
Create cutscene.lua

Zahir-Khan commented on map_gen/maps/crash_site/cutscene.lua:
```diff
@@ -0,0 +1,287 @@
+        function(event)
+            Task.set_timeout_in_ticks(60, start_cutscene, {event = event})
+        end
+)
```
This code Block is supposed to work, a similar implementation works on Diggy but not on Crashsite. I have tried hardcoding cutscene = true in config file. I have also tried RedmewConfig.player_create.cutscene = true in scenario file. 
Something is interfering in this particular scenario. Needs investigation. Right now the only way to enable cutscene is by using command /replay.

grilledham replied:
Maybe I'm being blind, but where is `CrashsiteCutscene.register()` called?

Zahir-Khan replied:
Oh, I completely forgot Diggy and Crash site have completely different code architecture. I thought .register was a factorio API I didn't know about. Fixed it.

Zahir-Khan committed:
Update cutscene.lua

Zahir-Khan committed:
Update cutscene.lua

Zahir-Khan committed:
Update scenario.lua

Zahir-Khan committed:
Update scenario.lua

Zahir-Khan committed:
Update cutscene.lua

minor addition

Zahir-Khan committed:
Update cutscene.lua

minor addition

theorangeangle committed:
small changes to improve grammar and flow

theorangeangle committed:
small changes to improve grammar and flow

grilledham commented:
Thanks, I'll review this at the weekend.

Zahir-Khan committed:
Added cutscene init

Zahir-Khan committed:
Added cutscene init

Zahir-Khan committed:
Timing tweaks for the new strings.

Zahir-Khan committed:
Timing tweaks for the new strings.

Zahir-Khan commented:
I have reviewed the timing for the new strings. Let me know if any more tweaking or changes is required. Go ahead and edit them if you like to.

Zahir-Khan committed:
Added player valid check

Zahir-Khan committed:
Added player valid check

Zahir-Khan committed:
Condensed code

Zahir-Khan committed:
Condensed code

Zahir-Khan commented:
#988 That needs to be closed too with this #1006.

grilledham commented:
When I run this branch locally with debug mode I get the following error when starting the scenario
![image](https://user-images.githubusercontent.com/10003572/69493507-59f86600-0ea7-11ea-9d8e-3f7f64e32197.png)

grilledham committed:
missing entity + desync fixes

grilledham committed:
missing entity + desync fixes

grilledham commented:
OK, I've pushed a commit that fixes the above issue and a desync issue I found when testing.

However there a re still some problems with this.
If you:
 - Start a server.
 - Join.
 - Leave while cutscene is active.
 - Rejoin.

You can't skip the cutsscene and more worryingly when the cutscnece ends you don't get returned to your character, you remain in spectator mode.

And a different minor issue, you have to press skip cutscene twice to skip both cutscenes, I'm not sure how I feel about that.

Zahir-Khan committed:
one skip btn press needed

Added supporting changes to cutscene_controller.lua

Zahir-Khan committed:
one skip btn press needed

Added supporting changes to cutscene_controller.lua

Zahir-Khan committed:
Prevent character collision

Zahir-Khan committed:
Prevent character collision

Zahir-Khan commented:
> OK, I've pushed a commit that fixes the above issue and a desync issue I found when testing.
> 
> However there a re still some problems with this.
> If you:
> 
> * Start a server.
> * Join.
> * Leave while cutscene is active.
> * Rejoin.
> 
> You can't skip the cutsscene and more worryingly when the cutscnece ends you don't get returned to your character, you remain in spectator mode.
> 
> And a different minor issue, you have to press skip cutscene twice to skip both cutscenes, I'm not sure how I feel about that.

Both issues have been fixed. Additional Improvements have been made. 
Important Note: Supporting changes need to be pushed to "cutscene_controller.lua". Holding out, on the account of #988 not being completed.

Zahir-Khan committed:
Merge and Update

Zahir-Khan committed:
Update cutscene.lua

Zahir-Khan committed:
Enhancements to enable multi-surface cutscenes

The setting of the skip_btn_flag to true will...
1. Enable one press to skip cutscene while in multiple.
2. Code paths updated to ensure the Player doesn't get locked out of the game while in a multi-surface cutscene.
3. Do different things depending on which cutscene was canceled in a multi-surface cutscene.

Zahir-Khan committed:
Update cutscene.lua

Zahir-Khan commented:
The setting of the skip_btn_flag to true will...
1. Enable one press to skip cutscene while in a multi-surface cutscene.
2. Code paths updated to ensure the Player doesn't get locked out of the game while in a multi-surface cutscene.
3. Do different things depending on which cutscene was canceled in a multi-surface cutscene.

grilledham commented on locale/en/redmew_maps.cfg:
```diff
@@ -65,6 +65,30 @@ cutscene_case7_line1=This concludes the introduction
+cutscene1_case_line6=The following introduction will help you get started playing!
+cutscene1_case1_line1=This is the starting area. It has a market and a coin mint.
+cutscene1_case1_line2=The coin mint produces coins at a limited rate, upgradable at the market.
+cutscene1_case1_line3=Coins can also be obtained by killing biters,spitters,worms and spawners.
```
No space after `,`s

grilledham commented on locale/en/redmew_maps.cfg:
```diff
@@ -65,6 +65,30 @@ cutscene_case7_line1=This concludes the introduction
+cutscene2_case_line2=You will find many structures similar to this dotted across the map.
+cutscene2_case0_line1=Outposts are defended by enemy turrets and walls. Turret strength scales with the rewards inside.
+cutscene2_case0_line2=To "capture" an outpost, you must destroy all turrets.
+cutscene2_case1_line1=The Outpost contains a Market, Loot Chests and Magic Crafters (assemblers or chemical plants).
```
Maybe `assemblers, chemical plants, etc`
It can also be electric furnaces and oil refineries.

grilledham commented on map_gen/maps/crash_site/cutscene.lua:
```diff
@@ -0,0 +1,328 @@
+
+        delayed_function(delayed_draw_arrow, player, tick, {settings = settings, offset = {x = 7, y = 2.5}, params = {rotation = rad(-45), time_to_live = 275 * 3 - 30}, fit_to_edge = true}, 0)
+
+        draw_text_auto_replacing(tick, settings, {x = 8.5, y = 3}, {'This is our toolbar!'}, player, {scale = 2.5, alignment = 'left'}, false, true, 275)
```
This message and the ones below could be added to the locale file.

grilledham reviewed (approved):
This is looking really good. Thanks for all the work you have done.

I think we can merge this, but you may wish to address the few comments I left

Zahir-Khan committed:
Update redmew_maps.cfg

Zahir-Khan committed:
Update cutscene.lua

Zahir-Khan committed:
Update redmew_maps.cfg

Zahir-Khan committed:
Update redmew_maps.cfg

Zahir-Khan committed:
Update cutscene.lua
